### ISSUE TITLE
Loss function ground truth processing

### ISSUE THREAD

**Issue by tutay** [1991viet/Yolo-pytorch#14]
Hi viet,
I need to ask you about something. In your loss function line 140-141.
why did you divide the ground truth w,h to best selected anchor width and height and take the logarithm.

`tcoord[b][best_n][2][gj * width + gi] = math.log(max(gt[i, 2], 1.0) / self.anchors[best_n, 0]) tcoord[b][best_n][3][gj * width + gi] = math.log(max(gt[i, 3], 1.0) / self.anchors[best_n, 1])
`
I cannot see this in the original paper. Most probably i missed something. Thanks in advance.

**Comment by tutay** [comment]
In the loss function line 56-59, prediction coordinates is changed to real coordinates with the below code snippet. 
` pred_boxes[:, 0] = (coord[:, :, 0].detach() + lin_x).view(-1)`
`pred_boxes[:, 1] = (coord[:, :, 1].detach() + lin_y).view(-1)`
`pred_boxes[:, 2] = (coord[:, :, 2].detach().exp() * anchor_w).view(-1)`
`pred_boxes[:, 3] = (coord[:, :, 3].detach().exp() * anchor_h).view(-1)`
Center x and y coordinates is shifted wrt the grid locations  and predicted w,h is processed with the formula given in the original paper. Predicted coordinates are changed to the real coordinates, aren't they?

If we apply below transformation, we are doing it twice I guess.
`tcoord[b][best_n][2][gj * width + gi] = math.log(max(gt[i, 2], 1.0) / self.anchors[best_n, 0]) tcoord[b][best_n][3][gj * width + gi] = math.log(max(gt[i, 3], 1.0) / self.anchors[best_n, 1]) `

Do we have to process the ground truth locations again to create the loss function.  I lossed in this two transformation. Prediction coordinates are changed to the real coordinates and ground truth is changed to prediction coordinates.  I thought, we just need to do one of above cases. 

Can you clarify this?
thank in advance.

**Comment by tutay** [comment]
I get the idea here. Below code is not used in directly in loss function calculation. 
`pred_boxes[:, 0] = (coord[:, :, 0].detach() + lin_x).view(-1)`
`pred_boxes[:, 1] = (coord[:, :, 1].detach() + lin_y).view(-1)`
`pred_boxes[:, 2] = (coord[:, :, 2].detach().exp() * anchor_w).view(-1)`
`pred_boxes[:, 3] = (coord[:, :, 3].detach().exp() * anchor_h).view(-1)`  
Instead ground truth is transformed to the what the network calculates. Below section converts the true locations to what network calculates.
`tcoord[b][best_n][2][gj * width + gi] = math.log(max(gt[i, 2], 1.0) / self.anchors[best_n, 0])`
`tcoord[b][best_n][3][gj * width + gi] = math.log(max(gt[i, 3], 1.0) / self.anchors[best_n, 1])`
